diff --git a/src/scripting.c b/src/scripting.c
index 53c0c9e..f993eaa 100644
--- a/src/scripting.c
+++ b/src/scripting.c
@@ -38,6 +38,9 @@
 #include <ctype.h>
 #include <math.h>
 
+#define lua_strlen(L,i)		lua_rawlen(L, (i))
+#define lua_open()	luaL_newstate()
+
 char *redisProtocolToLuaType_Int(lua_State *lua, char *reply);
 char *redisProtocolToLuaType_Bulk(lua_State *lua, char *reply);
 char *redisProtocolToLuaType_Status(lua_State *lua, char *reply);
