diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 39d297a..fec2ddf 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -114,6 +114,9 @@ find_package (PETSc COMPONENTS CXX REQUIRED)
 
 if (EXISTS "${PETSC_DIR}/${PETSC_ARCH}/conf/PETScConfig.cmake")
   include("${PETSC_DIR}/${PETSC_ARCH}/conf/PETScConfig.cmake")
+elseif (EXISTS "${PETSC_INCLUDES}/${PETSC_ARCH}/conf/PETScConfig.cmake")
+  include("${PETSC_INCLUDES}/${PETSC_ARCH}/conf/PETScConfig.cmake")
+  include_directories("${PETSC_INCLUDES}/${PETSC_ARCH}/include")
 elseif (EXISTS "${PETSC_DIR}/${PETSC_ARCH}/lib/petsc/conf/PETScConfig.cmake")
   include("${PETSC_DIR}/${PETSC_ARCH}/lib/petsc/conf/PETScConfig.cmake")
 else()
diff --git a/src/applications/dynamic_simulation_implicit/ds_app.hpp b/src/applications/dynamic_simulation_implicit/ds_app.hpp
index 40b63f2..a0298e4 100644
--- a/src/applications/dynamic_simulation_implicit/ds_app.hpp
+++ b/src/applications/dynamic_simulation_implicit/ds_app.hpp
@@ -18,7 +18,8 @@
 #ifndef _ds_app_h_
 #define _ds_app_h_
 
-#include "boost/smart_ptr/shared_ptr.hpp"
+#include <boost/serialization/type_info_implementation.hpp>
+#include <boost/smart_ptr/shared_ptr.hpp>
 
 namespace gridpack {
 namespace dsimplicit {
diff --git a/src/applications/dynamic_simulation_implicit/ds_factory.cpp b/src/applications/dynamic_simulation_implicit/ds_factory.cpp
index 37d0de7..6149e48 100644
--- a/src/applications/dynamic_simulation_implicit/ds_factory.cpp
+++ b/src/applications/dynamic_simulation_implicit/ds_factory.cpp
@@ -15,7 +15,8 @@
  */
 // -------------------------------------------------------------
 
-#include "boost/smart_ptr/shared_ptr.hpp"
+#include <boost/serialization/type_info_implementation.hpp>
+#include <boost/smart_ptr/shared_ptr.hpp>
 #include "gridpack/factory/base_factory.hpp"
 #include "ds_factory.hpp"
 #include "ds_components.hpp"
diff --git a/src/applications/examples/hello_world/CMakeLists.txt b/src/applications/examples/hello_world/CMakeLists.txt
index 0badd7f..66f730c 100644
--- a/src/applications/examples/hello_world/CMakeLists.txt
+++ b/src/applications/examples/hello_world/CMakeLists.txt
@@ -15,6 +15,7 @@
 
 set(target_libraries
     gridpack_components
+    gridpack_configuration
     gridpack_factory
     gridpack_partition
     gridpack_parallel
@@ -72,6 +73,6 @@ install(FILES
   ${CMAKE_CURRENT_SOURCE_DIR}/hw_factory.hpp
   ${CMAKE_CURRENT_SOURCE_DIR}/hw_main.cpp
   ${CMAKE_CURRENT_SOURCE_DIR}/10x10.raw
-  DESTINATION doc/gridpack/example/hello
+  DESTINATION share/doc/gridpack/example/hello
 )
 
diff --git a/src/applications/examples/hello_world/hw_app.cpp b/src/applications/examples/hello_world/hw_app.cpp
index e21c475..2762b81 100644
--- a/src/applications/examples/hello_world/hw_app.cpp
+++ b/src/applications/examples/hello_world/hw_app.cpp
@@ -16,7 +16,8 @@
 // -------------------------------------------------------------
 
 #include <iostream>
-#include "boost/smart_ptr/shared_ptr.hpp"
+#include <boost/serialization/type_info_implementation.hpp>
+#include <boost/smart_ptr/shared_ptr.hpp>
 #include "hw_app.hpp"
 #include "hw_factory.hpp"
 #include "gridpack/parser/PTI23_parser.hpp"
diff --git a/src/applications/examples/resistor_grid/CMakeLists.txt b/src/applications/examples/resistor_grid/CMakeLists.txt
index 6118773..271bf7a 100644
--- a/src/applications/examples/resistor_grid/CMakeLists.txt
+++ b/src/applications/examples/resistor_grid/CMakeLists.txt
@@ -81,5 +81,5 @@ install(FILES
   ${CMAKE_CURRENT_SOURCE_DIR}/rg_grid.F90
   ${CMAKE_CURRENT_BINARY_DIR}/input.xml
   ${CMAKE_CURRENT_BINARY_DIR}/small.raw
-  DESTINATION doc/gridpack/example/resistor_grid
+  DESTINATION share/doc/gridpack/example/resistor_grid
 )
diff --git a/src/applications/powerflow/CMakeLists.txt b/src/applications/powerflow/CMakeLists.txt
index 259335d..8dbd492 100644
--- a/src/applications/powerflow/CMakeLists.txt
+++ b/src/applications/powerflow/CMakeLists.txt
@@ -116,6 +116,6 @@ install(FILES
   ${CMAKE_CURRENT_SOURCE_DIR}/gridpack.petscrc
   ${CMAKE_CURRENT_SOURCE_DIR}/118.xml
   ${CMAKE_CURRENT_SOURCE_DIR}/118.raw
-  DESTINATION doc/gridpack/example/powerflow
+  DESTINATION share/doc/gridpack/example/powerflow
 )
 
diff --git a/src/applications/powerflow2/CMakeLists.txt b/src/applications/powerflow2/CMakeLists.txt
index 9ac46b0..9345d43 100644
--- a/src/applications/powerflow2/CMakeLists.txt
+++ b/src/applications/powerflow2/CMakeLists.txt
@@ -101,5 +101,5 @@ install(FILES
   ${CMAKE_CURRENT_SOURCE_DIR}/IEEE14.raw
   ${CMAKE_CURRENT_SOURCE_DIR}/118.xml
   ${CMAKE_CURRENT_SOURCE_DIR}/118.raw
-  DESTINATION doc/gridpack/example/powerflow2
+  DESTINATION share/doc/gridpack/example/powerflow2
 )
diff --git a/src/mapper/CMakeLists.txt b/src/mapper/CMakeLists.txt
index 92f7da5..348928a 100644
--- a/src/mapper/CMakeLists.txt
+++ b/src/mapper/CMakeLists.txt
@@ -15,6 +15,7 @@
 
 set(target_libraries
   gridpack_timer
+  gridpack_configuration
   ${GA_LIBRARIES} 
   ${Boost_LIBRARIES} 
   ${MPI_CXX_LIBRARIES}
diff --git a/src/optimization/expression.cpp b/src/optimization/expression.cpp
index 36c1b58..322e874 100644
--- a/src/optimization/expression.cpp
+++ b/src/optimization/expression.cpp
@@ -16,6 +16,7 @@
 
 // Cannot do this because these classes have a boost::shared_ptr member
 // #include <boost/mpi.hpp>
+#include <boost/serialization/type_info_implementation.hpp>
 #include <boost/serialization/string.hpp>
 #include <boost/serialization/shared_ptr.hpp>
 #include <boost/serialization/weak_ptr.hpp>
diff --git a/src/optimization/optimizer.cpp b/src/optimization/optimizer.cpp
index d2d6119..22dd7c5 100644
--- a/src/optimization/optimizer.cpp
+++ b/src/optimization/optimizer.cpp
@@ -23,6 +23,7 @@
 
 // These two includes are needed for Boost 1.56
 
+#include <boost/serialization/type_info_implementation.hpp>
 #include <boost/serialization/singleton.hpp>
 #include <boost/serialization/extended_type_info.hpp>
 
diff --git a/src/optimization/optimizer.hpp b/src/optimization/optimizer.hpp
index bfb4d0b..711661e 100644
--- a/src/optimization/optimizer.hpp
+++ b/src/optimization/optimizer.hpp
@@ -22,6 +22,7 @@
 #define _optimizer_hpp_
 
 #include <vector>
+#include <map>
 #include <boost/scoped_ptr.hpp>
 #include <gridpack/utilities/uncopyable.hpp>
 #include <gridpack/utilities/exception.hpp>
diff --git a/src/optimization/test/expression_test.cpp b/src/optimization/test/expression_test.cpp
index 8ae7594..71fd2ad 100644
--- a/src/optimization/test/expression_test.cpp
+++ b/src/optimization/test/expression_test.cpp
@@ -26,6 +26,7 @@
 
 // These two includes are needed for Boost 1.56
 
+#include <boost/serialization/type_info_implementation.hpp>
 #include <boost/serialization/singleton.hpp>
 #include <boost/serialization/extended_type_info.hpp>
 
diff --git a/src/optimization/test/variable_test.cpp b/src/optimization/test/variable_test.cpp
index ee32ee0..34502a5 100644
--- a/src/optimization/test/variable_test.cpp
+++ b/src/optimization/test/variable_test.cpp
@@ -24,6 +24,7 @@
 #include <boost/bind.hpp>
 
 // These two includes are needed for Boost 1.56
+#include <boost/serialization/type_info_implementation.hpp>
 #include <boost/serialization/singleton.hpp>
 #include <boost/serialization/extended_type_info.hpp>
 
diff --git a/src/serial_io/CMakeLists.txt b/src/serial_io/CMakeLists.txt
index a9cdffd..ac5592f 100644
--- a/src/serial_io/CMakeLists.txt
+++ b/src/serial_io/CMakeLists.txt
@@ -18,6 +18,7 @@ set(target_libraries
   ${GA_LIBRARIES} 
   ${Boost_LIBRARIES} 
   ${MPI_CXX_LIBRARIES}
+  ${PETSC_LIBRARIES}
 )
 
 if (GOSS_DIR)
