diff --git a/CMakeLists.txt b/CMakeLists.txt
index 27e4890..a0dc090 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -405,8 +405,8 @@ install(FILES
   )
 
 if(EIGEN_BUILD_PKGCONFIG)
-    configure_file(eigen3.pc.in eigen3.pc @ONLY)
-    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/eigen3.pc
+    configure_file(eigen3.pc.in eigen3-tensorflow.pc @ONLY)
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/eigen3-tensorflow.pc
         DESTINATION ${PKGCONFIG_INSTALL_DIR}
         )
 endif()
@@ -525,11 +525,11 @@ if (NOT CMAKE_VERSION VERSION_LESS 3.0)
   # Export as title case Eigen
   set_target_properties (eigen PROPERTIES EXPORT_NAME Eigen)
 
-  install (TARGETS eigen EXPORT Eigen3Targets)
+  install (TARGETS eigen EXPORT Eigen3TensorflowTargets)
 
   configure_package_config_file (
     ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Eigen3Config.cmake.in
-    ${CMAKE_CURRENT_BINARY_DIR}/Eigen3Config.cmake
+    ${CMAKE_CURRENT_BINARY_DIR}/Eigen3TensorflowConfig.cmake
     PATH_VARS EIGEN_INCLUDE_DIR EIGEN_ROOT_DIR
     INSTALL_DESTINATION ${CMAKEPACKAGE_INSTALL_DIR}
     NO_CHECK_REQUIRED_COMPONENTS_MACRO # Eigen does not provide components
@@ -540,19 +540,19 @@ if (NOT CMAKE_VERSION VERSION_LESS 3.0)
   # used for 32 bit targets as well (and vice versa).
   set (_Eigen3_CMAKE_SIZEOF_VOID_P ${CMAKE_SIZEOF_VOID_P})
   unset (CMAKE_SIZEOF_VOID_P)
-  write_basic_package_version_file (Eigen3ConfigVersion.cmake
+  write_basic_package_version_file (Eigen3TensorflowConfigVersion.cmake
                                     VERSION ${EIGEN_VERSION_NUMBER}
                                     COMPATIBILITY SameMajorVersion)
   set (CMAKE_SIZEOF_VOID_P ${_Eigen3_CMAKE_SIZEOF_VOID_P})
 
   # The Eigen target will be located in the Eigen3 namespace. Other CMake
   # targets can refer to it using Eigen3::Eigen.
-  export (TARGETS eigen NAMESPACE Eigen3:: FILE Eigen3Targets.cmake)
+  export (TARGETS eigen NAMESPACE Eigen3:: FILE Eigen3TensorflowTargets.cmake)
   # Export Eigen3 package to CMake registry such that it can be easily found by
   # CMake even if it has not been installed to a standard directory.
   export (PACKAGE Eigen3)
 
-  install (EXPORT Eigen3Targets NAMESPACE Eigen3:: DESTINATION ${CMAKEPACKAGE_INSTALL_DIR})
+  install (EXPORT Eigen3TensorflowTargets NAMESPACE Eigen3:: DESTINATION ${CMAKEPACKAGE_INSTALL_DIR})
 
 else (NOT CMAKE_VERSION VERSION_LESS 3.0)
   # Fallback to legacy Eigen3Config.cmake without the imported target
@@ -564,7 +564,7 @@ else (NOT CMAKE_VERSION VERSION_LESS 3.0)
   if(CPCH_PATH)
     configure_package_config_file (
       ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Eigen3ConfigLegacy.cmake.in
-      ${CMAKE_CURRENT_BINARY_DIR}/Eigen3Config.cmake
+      ${CMAKE_CURRENT_BINARY_DIR}/Eigen3TensorflowConfig.cmake
       PATH_VARS EIGEN_INCLUDE_DIR EIGEN_ROOT_DIR
       INSTALL_DESTINATION ${CMAKEPACKAGE_INSTALL_DIR}
       NO_CHECK_REQUIRED_COMPONENTS_MACRO # Eigen does not provide components
@@ -576,19 +576,19 @@ else (NOT CMAKE_VERSION VERSION_LESS 3.0)
     set(PACKAGE_EIGEN_INCLUDE_DIR ${EIGEN_INCLUDE_DIR})
     set(PACKAGE_EIGEN_ROOT_DIR ${EIGEN_ROOT_DIR})
     configure_file ( ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Eigen3ConfigLegacy.cmake.in
-                     ${CMAKE_CURRENT_BINARY_DIR}/Eigen3Config.cmake
+                     ${CMAKE_CURRENT_BINARY_DIR}/Eigen3TensorflowConfig.cmake
                      @ONLY ESCAPE_QUOTES )
   endif()
 
-  write_basic_package_version_file( Eigen3ConfigVersion.cmake
+  write_basic_package_version_file( Eigen3TensorflowConfigVersion.cmake
                                     VERSION ${EIGEN_VERSION_NUMBER}
                                     COMPATIBILITY SameMajorVersion )
 
 endif (NOT CMAKE_VERSION VERSION_LESS 3.0)
 
 install ( FILES ${CMAKE_CURRENT_SOURCE_DIR}/cmake/UseEigen3.cmake
-                ${CMAKE_CURRENT_BINARY_DIR}/Eigen3Config.cmake
-                ${CMAKE_CURRENT_BINARY_DIR}/Eigen3ConfigVersion.cmake
+                ${CMAKE_CURRENT_BINARY_DIR}/Eigen3TensorflowConfig.cmake
+                ${CMAKE_CURRENT_BINARY_DIR}/Eigen3TensorflowConfigVersion.cmake
           DESTINATION ${CMAKEPACKAGE_INSTALL_DIR} )
 
 # Add uninstall target
diff --git a/cmake/Eigen3Config.cmake.in b/cmake/Eigen3Config.cmake.in
index c5c5468..39428f0 100644
--- a/cmake/Eigen3Config.cmake.in
+++ b/cmake/Eigen3Config.cmake.in
@@ -3,7 +3,7 @@
 
 @PACKAGE_INIT@
 
-include ("${CMAKE_CURRENT_LIST_DIR}/Eigen3Targets.cmake")
+include ("${CMAKE_CURRENT_LIST_DIR}/Eigen3TensorflowTargets.cmake")
 
 # Legacy variables, do *not* use. May be removed in the future.
 
