diff --git a/lib_pypy/_gdbm_build.py b/lib_pypy/_gdbm_build.py
index 8578874..80d439c 100644
--- a/lib_pypy/_gdbm_build.py
+++ b/lib_pypy/_gdbm_build.py
@@ -11,7 +11,7 @@ ffi.cdef('''
 #define GDBM_NOLOCK ...
 #define GDBM_REPLACE ...
 
-void* gdbm_open(char *, int, int, int, void (*)());
+void* gdbm_open(char *, int, int, int, void (*)(const char *));
 void gdbm_close(void*);
 
 typedef struct {
