*** etc/profile	Tue Sep 15 14:54:13 2015
--- etc/profile	Thu Sep 17 15:44:15 2015
*************** if [[ -n "${_ro_root}" ]]; then
*** 58,69 ****
--- 58,84 ----
  	done
  fi

+ if [[ -z "$USING_ESELECT_ENVMOD" ]]
+ 	export INFOPATH_BAK="$INFOPATH"
+ 	export MANPATH_BAK="$MANPATH"
+ 	export PATH_BAK="$PATH"
+ 	[[ "$USER" == "" ]] && export USER="user"
+ fi
+ 
  # Load environment settings from profile.env, which is created by
  # env-update from the files in "@GENTOO_PORTAGE_EPREFIX@"/etc/env.d
  if [ -e "@GENTOO_PORTAGE_EPREFIX@"/etc/profile.env ] ; then
  	. "@GENTOO_PORTAGE_EPREFIX@"/etc/profile.env
  fi
  
+ if [[ -z "$USING_ESELECT_ENVMOD" ]]
+ 	export INFOPATH="$INFOPATH:$INFOPATH_BAK"
+ 	export MANPATH="$MANPATH:$MANPATH_BAK"
+ 	export PATH="$PATH:$PATH_BAK"
+ 	[[ -f /etc/profile ]] && . /etc/profile >/dev/null 2>&1
+ fi
+ unset INFOPATH_BAK MANPATH_BAK PATH_BAK
+ 
  # You should override these in your ~/.bashrc (or equivalent) for per-user
  # settings.  For system defaults, you can add a new file in "@GENTOO_PORTAGE_EPREFIX@"/etc/profile.d/.
  export EDITOR=${EDITOR:-"@GENTOO_PORTAGE_EPREFIX@"/bin/nano}
*************** elif [[ "${EPREFIX}" != "@GENTOO_PORTAGE
*** 171,176 ****
--- 186,219 ----
  	echo "         Having EPREFIX set this way will probably render this environment unusable."
  fi
  
+ if [[ ${_ro_recursion_level} == 0 ]]
+ then
+ 	if [[ -z "$USING_ESELECT_ENVMOD" ]]
+ 	then
+ 		#
+ 		# Load the necessary Cray modules
+ 		#
+ 		#module switch cray-mpich cray-mpich/7.2.4
+ 		echo "Loading default environment modules. Use app-eselect/eselect-cray to change."
+ 		eval "$(modulecmd bash load craype gcc)"
+ 		eval "$(modulecmd bash switch PrgEnv-cray PrgEnv-gnu)" #>/dev/null 2>&1 #Suppress warning if module has already been switched
+ 		eval "$(modulecmd bash load cmake cray-hdf5 cray-netcdf cray-tpsl fftw)"
+ 		eval "$(modulecmd bash unload xalt)"
+ 
+ 		LD_LIB_PATHS=( $(echo $LD_LIBRARY_PATH | tr ':' ' ') )
+ 		for path in ${LD_LIB_PATHS[@]} ; do
+ 			LDP_LDFLAGS="$LDP_LDFLAGS -Wl,--rpath=$path"
+ 		done
+
+ 		export LDP_LDFLAGS
+ 	fi
+ 
+ 	PATH="$(echo $PATH | sed 's|:/usr/bin:/bin||g'):/usr/bin:/bin"
+ 
+ 	export CRAY_PKG_CONFIG_PATH="$(cc --cray-print-opts=pkg_config_path)"
+ 
+ 	export PKG_CONFIG_PATH="@GENTOO_PORTAGE_EPREFIX@/usr/lib/pkgconfig:$CRAY_PKG_CONFIG_PATH:$PKG_CONFIG_PATH:/usr/lib64/pkgconfig"
+ fi
+ 
  unset _ro_root
  unset _ro_deps
  unset _ro_chained_path_vars
