diff --git a/CMake/FindMySQL.cmake b/CMake/FindMySQL.cmake
index d021682..e2eab11 100644
--- a/CMake/FindMySQL.cmake
+++ b/CMake/FindMySQL.cmake
@@ -18,6 +18,10 @@
 SET( MYSQL_FOUND 0 )
 
 FIND_PATH(MYSQL_INCLUDE_DIRECTORIES mysql.h
+  ${MYSQL_ROOT}/include
+  ${MYSQL_ROOT}/include/mysql
+  $ENV{EPREFIX}/usr/include
+  $ENV{EPREFIX}/usr/include/mysql
   /usr/include
   /usr/include/mysql
   /usr/local/include
@@ -34,6 +38,8 @@ FIND_PATH(MYSQL_INCLUDE_DIRECTORIES mysql.h
 FIND_LIBRARY( MYSQL_LIBRARY
   NAMES mysql libmysql mysqlclient
   PATHS
+  ${MYSQL_ROOT}/lib
+  $ENV{EPREFIX}/usr/lib
   /usr/lib
   /usr/lib/mysql
   /usr/local/lib
@@ -51,10 +57,8 @@ FIND_LIBRARY( MYSQL_LIBRARY
 # to build MYSQL stuff.
 
 IF (NOT WIN32)
-  FIND_LIBRARY( MYSQL_EXTRA_LIBRARIES
-                NAMES z zlib
-                PATHS /usr/lib /usr/local/lib
-                DOC "If more libraries are necessary to link in a MySQL client (typically zlib), specify them here.")
+  FIND_PACKAGE(ZLIB REQUIRED)
+  SET(MYSQLV_EXTRA_LIBRARIES ${ZLIB_LIBRARIES})
 ELSE ()
   SET( MYSQL_EXTRA_LIBRARIES "" )
 ENDIF ()
diff --git a/ThirdParty/AutobahnPython/module.cmake b/ThirdParty/AutobahnPython/module.cmake
index a5dd9ef..ee12ce5 100644
--- a/ThirdParty/AutobahnPython/module.cmake
+++ b/ThirdParty/AutobahnPython/module.cmake
@@ -1,8 +1,6 @@
 vtk_module(AutobahnPython
   DEPENDS
-    SixPython
     Twisted
-    txaio
     vtkPython
   OPTIONAL_PYTHON_LINK
   EXCLUDE_FROM_WRAPPING)
diff --git a/ThirdParty/Twisted/module.cmake b/ThirdParty/Twisted/module.cmake
index a15e447..ee4ed31 100644
--- a/ThirdParty/Twisted/module.cmake
+++ b/ThirdParty/Twisted/module.cmake
@@ -1,9 +1,4 @@
 vtk_module(Twisted
-  DEPENDS
-    constantly
-    hyperlink
-    incremental
-    ZopeInterface
   COMPILE_DEPENDS
     vtkPython
   OPTIONAL_PYTHON_LINK
diff --git a/CMake/FindLIBPROJ4.cmake b/CMake/FindLIBPROJ4.cmake
index 9d3904d..a76e89d 100644
--- a/CMake/FindLIBPROJ4.cmake
+++ b/CMake/FindLIBPROJ4.cmake
@@ -15,16 +15,6 @@ if ( NOT LIBPROJ4_INCLUDE_DIR OR NOT LIBPROJ4_LIBRARIES OR NOT LIBPROJ4_FOUND )
     NAMES proj
     PATHS
       ${_LIBPROJ4_DIR}/lib64
-      ${CMAKE_INSTALL_PREFIX}/lib64
-      /usr/local/lib64
-      /usr/lib64
-      ${_LIBPROJ4_DIR}
-      ${_LIBPROJ4_DIR}/lib
-      ${CMAKE_INSTALL_PREFIX}/bin
-      ${CMAKE_INSTALL_PREFIX}/lib
-      /usr/local/lib
-      /usr/lib
-    NO_DEFAULT_PATH
   )
 
   find_path( LIBPROJ4_INCLUDE_DIR
@@ -32,12 +22,6 @@ if ( NOT LIBPROJ4_INCLUDE_DIR OR NOT LIBPROJ4_LIBRARIES OR NOT LIBPROJ4_FOUND )
     PATHS
       ${_LIBPROJ4_DIR}
       ${_LIBPROJ4_DIR}/include
-      ${CMAKE_INSTALL_PREFIX}/include
-      /usr/local/pgsql/include
-      /usr/local/include
-      /usr/include
-      /usr/include/postgresql
-    NO_DEFAULT_PATH
   )
 
   if ( NOT LIBPROJ4_INCLUDE_DIR OR NOT LIBPROJ4_LIBRARIES )
diff --git a/CMake/vtkModuleMacros.cmake b/CMake/vtkModuleMacros.cmake
index ffbc0ca..d945ad6 100644
--- a/CMake/vtkModuleMacros.cmake
+++ b/CMake/vtkModuleMacros.cmake
@@ -77,7 +77,7 @@ macro(vtk_module _name)
   foreach(arg ${ARGN})
     # XXX: Adding a new keyword? Update Utilities/Maintenance/WhatModulesVTK.py
     # and Utilities/Maintenance/VisualizeModuleDependencies.py as well.
-    if("${arg}" MATCHES "^((|COMPILE_|PRIVATE_|TEST_|)DEPENDS|DESCRIPTION|TCL_NAME|IMPLEMENTS|BACKEND|DEFAULT|GROUPS|TEST_LABELS|KIT|LEGACY)$")
+    if("${arg}" MATCHES "^((|COMPILE_|PRIVATE_|INTERFACE_|TEST_|)DEPENDS|DESCRIPTION|TCL_NAME|IMPLEMENTS|BACKEND|DEFAULT|GROUPS|TEST_LABELS|KIT|LEGACY)$")
       set(_doing "${arg}")
     elseif("${arg}" STREQUAL "EXCLUDE_FROM_ALL")
       set(_doing "")
@@ -111,6 +111,8 @@ macro(vtk_module _name)
       list(APPEND ${vtk-module}_COMPILE_DEPENDS "${arg}")
     elseif("${_doing}" STREQUAL "PRIVATE_DEPENDS")
       list(APPEND ${vtk-module}_PRIVATE_DEPENDS "${arg}")
+    elseif("${_doing}" STREQUAL "INTERFACE_DEPENDS")
+      list(APPEND ${vtk-module}_INTERFACE_DEPENDS "${arg}")
     elseif("${_doing}" STREQUAL "DESCRIPTION")
       set(_doing "")
       set(${vtk-module}_DESCRIPTION "${arg}")
@@ -157,10 +159,12 @@ macro(vtk_module _name)
   set(${vtk-module}_LINK_DEPENDS "${${vtk-module}_DEPENDS}")
   set(${vtk-module}_WRAP_DEPENDS
     ${${vtk-module}_DEPENDS}
-    ${${vtk-module}_PRIVATE_DEPENDS})
+    ${${vtk-module}_PRIVATE_DEPENDS}
+    ${${vtk-module}_INTERFACE_DEPENDS})
   list(APPEND ${vtk-module}_DEPENDS
     ${${vtk-module}_COMPILE_DEPENDS}
-    ${${vtk-module}_PRIVATE_DEPENDS})
+    ${${vtk-module}_PRIVATE_DEPENDS}
+    ${${vtk-module}_INTERFACE_DEPENDS})
   unset(${vtk-module}_COMPILE_DEPENDS)
   list(SORT ${vtk-module}_DEPENDS) # Deterministic order.
   list(SORT ${vtk-module-test}_DEPENDS) # Deterministic order.
@@ -750,6 +754,16 @@ function(vtk_module_library name)
     vtk_module_link_libraries(${vtk-module} LINK_PRIVATE ${${dep}_LIBRARIES})
   endforeach()
 
+  foreach(dep IN LISTS ${vtk-module}_INTERFACE_DEPENDS)
+    if(${dep}_INCLUDE_DIRS)
+      include_directories(${${dep}_INCLUDE_DIRS})
+    endif()
+    #if(${dep}_LIBRARY_DIRS)
+    #  link_directories(${${dep}_LIBRARY_DIRS})
+    #endif()
+    target_link_libraries(${vtk-module} INTERFACE ${${dep}_LIBRARIES})
+  endforeach()
+
   if(${vtk-module}_LEGACY)
     set(legacy_message "")
     string(APPEND legacy_message ${vtk-module} " module was deprecated for VTK "
diff --git a/Filters/Python/module.cmake b/Filters/Python/module.cmake
index 4d88930..fe85a1d 100644
--- a/Filters/Python/module.cmake
+++ b/Filters/Python/module.cmake
@@ -15,6 +15,7 @@ if (VTK_WRAP_PYTHON)
       vtkCommonExecutionModel
     PRIVATE_DEPENDS
       vtkCommonCore
+    INTERFACE_DEPENDS
       vtkWrappingPythonCore
     )
 endif ()
diff --git a/Parallel/MPI4Py/module.cmake b/Parallel/MPI4Py/module.cmake
index b4f01ad..4236955 100644
--- a/Parallel/MPI4Py/module.cmake
+++ b/Parallel/MPI4Py/module.cmake
@@ -14,8 +14,9 @@ if (VTK_WRAP_PYTHON)
       vtkWrapping
     DEPENDS
       vtkCommonCore
-      vtkPython
     PRIVATE_DEPENDS
       vtkParallelMPI
+    INTERFACE_DEPENDS
+      vtkPython
     )
-endif ()
\ No newline at end of file
+endif ()
diff --git a/Filters/Python/CMakeLists.txt b/Filters/Python/CMakeLists.txt
index 674a753..fde75f9 100644
--- a/Filters/Python/CMakeLists.txt
+++ b/Filters/Python/CMakeLists.txt
@@ -2,4 +2,4 @@ set(Module_SRCS
   vtkPythonAlgorithm.cxx
   )
 
-vtk_module_library(vtkFiltersPython ${Module_SRCS})
+vtk_module_library(vtkFiltersPythonPy${PYTHON_VERSION_MAJOR} ${Module_SRCS})
diff --git a/Filters/Python/module.cmake b/Filters/Python/module.cmake
index fe85a1d..5559a8b 100644
--- a/Filters/Python/module.cmake
+++ b/Filters/Python/module.cmake
@@ -1,5 +1,5 @@
 if (VTK_WRAP_PYTHON)
-  vtk_module(vtkFiltersPython
+  vtk_module(vtkFiltersPythonPy${PYTHON_VERSION_MAJOR}
     GROUPS
       StandAlone
     COMPILE_DEPENDS
diff --git a/Parallel/MPI4Py/CMakeLists.txt b/Parallel/MPI4Py/CMakeLists.txt
index 7fdebba..3215c4d 100644
--- a/Parallel/MPI4Py/CMakeLists.txt
+++ b/Parallel/MPI4Py/CMakeLists.txt
@@ -1,5 +1,5 @@
 include(vtkMPI)
-set(vtkParallelMPI4Py_SYSTEM_INCLUDE_DIRS ${MPI_C_INCLUDE_PATH})
+set(vtkParallelMPI4PyPy${PYTHON_VERSION_MAJOR}_SYSTEM_INCLUDE_DIRS ${MPI_C_INCLUDE_PATH})
 
 set(Module_SRCS
   vtkMPI4PyCommunicator.cxx
@@ -28,5 +28,5 @@ add_definitions("-DMPICH_IGNORE_CXX_SEEK")
 # FIXME: The ifdefs need to be removed, for now defining it to on
 set(${vtk-module}_EXPORT_CODE "#define VTK_USE_MPI")
 
-vtk_module_library(vtkParallelMPI4Py ${Module_SRCS})
-vtk_mpi_link(vtkParallelMPI4Py)
+vtk_module_library(vtkParallelMPI4PyPy${PYTHON_VERSION_MAJOR} ${Module_SRCS})
+vtk_mpi_link(vtkParallelMPI4PyPy${PYTHON_VERSION_MAJOR})
diff --git a/Parallel/MPI4Py/module.cmake b/Parallel/MPI4Py/module.cmake
index 4236955..2a8db8b 100644
--- a/Parallel/MPI4Py/module.cmake
+++ b/Parallel/MPI4Py/module.cmake
@@ -1,5 +1,5 @@
 if (VTK_WRAP_PYTHON)
-  vtk_module(vtkParallelMPI4Py
+  vtk_module(vtkParallelMPI4PyPy${PYTHON_VERSION_MAJOR}
     GROUPS
       MPI
     DEPENDS
diff --git a/Utilities/PythonInterpreter/CMakeLists.txt b/Utilities/PythonInterpreter/CMakeLists.txt
index 77a3b89..9dc5a12 100644
--- a/Utilities/PythonInterpreter/CMakeLists.txt
+++ b/Utilities/PythonInterpreter/CMakeLists.txt
@@ -6,5 +6,5 @@ set (Module_SRCS
 set (${vtk-module}_HDRS
   vtkPythonStdStreamCaptureHelper.h)
 
-vtk_module_library(vtkPythonInterpreter
+vtk_module_library(vtkPythonInterpreterPy${PYTHON_VERSION_MAJOR}
   ${Module_SRCS})
diff --git a/Utilities/PythonInterpreter/module.cmake b/Utilities/PythonInterpreter/module.cmake
index b014275..0ae7ba4 100644
--- a/Utilities/PythonInterpreter/module.cmake
+++ b/Utilities/PythonInterpreter/module.cmake
@@ -1,4 +1,4 @@
-vtk_module(vtkPythonInterpreter
+vtk_module(vtkPythonInterpreterPy${PYTHON_VERSION_MAJOR}
   KIT
     vtkWrapping
   DEPENDS
diff --git a/Rendering/Matplotlib/module.cmake b/Rendering/Matplotlib/module.cmake
index 042b6d5..a257559 100644
--- a/Rendering/Matplotlib/module.cmake
+++ b/Rendering/Matplotlib/module.cmake
@@ -5,7 +5,7 @@ elseif(VTK_RENDERING_BACKEND STREQUAL "OpenGL")
 elseif(VTK_RENDERING_BACKEND STREQUAL "OpenGL2")
   set(gl2ps_depends vtkRenderingGL2PSOpenGL2 vtkIOExportOpenGL2)
 endif()
-vtk_module(vtkRenderingMatplotlib
+vtk_module(vtkRenderingMatplotlibPy${PYTHON_VERSION_MAJOR}
   IMPLEMENTS
     vtkRenderingFreeType
   TEST_DEPENDS
@@ -21,14 +21,15 @@ vtk_module(vtkRenderingMatplotlib
     vtkViewsContext2D
     ${gl2ps_depends}
   DEPENDS
-    vtkPythonInterpreter
     vtkRenderingFreeType
   PRIVATE_DEPENDS
     vtkCommonCore
     vtkCommonDataModel
     vtkCommonTransforms
     vtkImagingCore
-    vtkPython
     vtkRenderingCore
+  INTERFACE_DEPENDS
     vtkWrappingPythonCore
-  )
\ No newline at end of file
+    vtkPythonInterpreterPy${PYTHON_VERSION_MAJOR}
+    vtkPython
+  )
