diff --git a/cmake/ConfigGen.cmake b/cmake/ConfigGen.cmake
index 4b447a6..82f8146 100644
--- a/cmake/ConfigGen.cmake
+++ b/cmake/ConfigGen.cmake
@@ -96,7 +96,7 @@ function(caffe_generate_export_configs)
   # remove source and build dir includes
   caffe_get_items_with_prefix(${PROJECT_SOURCE_DIR} Caffe_INCLUDE_DIRS __insource)
   caffe_get_items_with_prefix(${PROJECT_BINARY_DIR} Caffe_INCLUDE_DIRS __inbinary)
-  list(REMOVE_ITEM Caffe_INCLUDE_DIRS ${__insource} ${__inbinary})
+  list(REMOVE_ITEM Caffe_INCLUDE_DIRS __insource __inbinary)
 
   # add `install` include folder
   set(lines
diff --git a/cmake/Dependencies.cmake b/cmake/Dependencies.cmake
index ad8f1df..39c356e 100644
--- a/cmake/Dependencies.cmake
+++ b/cmake/Dependencies.cmake
@@ -26,7 +26,7 @@ include(cmake/ProtoBuf.cmake)
 # ---[ HDF5
 find_package(HDF5 COMPONENTS HL REQUIRED)
 include_directories(SYSTEM ${HDF5_INCLUDE_DIRS} ${HDF5_HL_INCLUDE_DIR})
-list(APPEND Caffe_LINKER_LIBS ${HDF5_LIBRARIES})
+list(APPEND Caffe_LINKER_LIBS ${HDF5_LIBRARIES} ${HDF5_HL_LIBRARIES})
 
 # ---[ LMDB
 if(USE_LMDB)
@@ -184,5 +184,5 @@ endif()
 if(NCCL_FOUND)
   add_definitions(-DUSE_NCCL)
   include_directories(SYSTEM ${NCCL_INCLUDE})
-  list(APPEND Caffe_LINKER_LIBS ${NCCL_LIBRARY})
+  list(APPEND Caffe_LINKER_LIBS ${NCCL_LIBRARY} -lrt -ldl)
 endif()
diff --git a/include/caffe/layers/detectnet_transform_layer.hpp b/include/caffe/layers/detectnet_transform_layer.hpp
index 4e41fc1..2a57925 100644
--- a/include/caffe/layers/detectnet_transform_layer.hpp
+++ b/include/caffe/layers/detectnet_transform_layer.hpp
@@ -35,7 +35,7 @@ struct AugmentSelection {
   cv::Point crop_offset;
 
   bool doHueRotation() const { return std::abs(hue_rotation) > FLT_EPSILON; }
-  bool doDesaturation() const { return saturation < (1.0 - 1.0/UINT8_MAX); }
+  bool doDesaturation() const { return saturation < (1.0 - 1.0/__UINT8_MAX__); }
   bool doScale(const cv::Size& s) const {
     return s.height != scale.height || s.width != scale.width;
   }
diff --git a/src/caffe/layers/detectnet_transform_layer.cpp b/src/caffe/layers/detectnet_transform_layer.cpp
index 1e13b99..50a4e92 100644
--- a/src/caffe/layers/detectnet_transform_layer.cpp
+++ b/src/caffe/layers/detectnet_transform_layer.cpp
@@ -83,7 +83,7 @@ void DetectNetTransformationLayer<Dtype>::retrieveMeanImage(Size dimensions) {
 #endif
 
   // scale from 0..255 to 0..1:
-  data_mean_ /= Dtype(UINT8_MAX);
+  data_mean_ /= Dtype(__UINT8_MAX__);
 }
 
 
@@ -91,12 +91,12 @@ template<typename Dtype>
 void DetectNetTransformationLayer<Dtype>::retrieveMeanChannels() {
   switch (t_param_.mean_value_size()) {
     case 1:
-      mean_values_.fill(t_param_.mean_value(0) / Dtype(UINT8_MAX));
+      mean_values_.fill(t_param_.mean_value(0) / Dtype(__UINT8_MAX__));
       break;
     case 3:
       for (size_t iChannel = 0; iChannel != 3; ++iChannel) {
         mean_values_[iChannel] =
-            t_param_.mean_value(iChannel) / Dtype(UINT8_MAX);
+            t_param_.mean_value(iChannel) / Dtype(__UINT8_MAX__);
       }
       break;
     case 0:
@@ -417,7 +417,7 @@ DetectNetTransformationLayer<Dtype>::transform_image_cpu(
     const Mat3v& src_img, const AugmentSelection& as
 ) {
   // Scale from [0,255] to [0,1] for HSV augmentations
-  Mat3v img = src_img.clone() / UINT8_MAX;
+  Mat3v img = src_img.clone() / __UINT8_MAX__;
 
   // Do HSV transformations before mean subtraction while image still in [0,1]
   if (as.doHueRotation() || as.doDesaturation()) {
@@ -441,7 +441,7 @@ DetectNetTransformationLayer<Dtype>::transform_image_cpu(
       as.crop_offset);
 
   // Scale from [-1,1] into [-255,255]
-  img *= Dtype(UINT8_MAX);
+  img *= Dtype(__UINT8_MAX__);
   return img;
 }
 
diff --git a/src/caffe/layers/detectnet_transform_layer.cu b/src/caffe/layers/detectnet_transform_layer.cu
index cf83335..7b55011 100644
--- a/src/caffe/layers/detectnet_transform_layer.cu
+++ b/src/caffe/layers/detectnet_transform_layer.cu
@@ -127,7 +127,7 @@ __global__ void color_transformations(
     // check what needs doing
     const AugmentSelection& as = aug_data[n];
     const bool doHueRotation = (abs(as.hue_rotation) > FLT_EPSILON);
-    const bool doDesaturation = (as.saturation < (1.0 - 1.0/UINT8_MAX));
+    const bool doDesaturation = (as.saturation < (1.0 - 1.0/__UINT8_MAX__));
 
     // N*cs*hs*ws + H*ws + W
     int index = n * shape.z * shape.y * shape.x + y * shape.x + x;
@@ -402,9 +402,9 @@ void DetectNetTransformationLayer<Dtype>::Forward_gpu(
     // NOLINT_NEXT_LINE(whitespace/operators)
     channel_mean_subtraction<<<CAFFE_GET_BLOCKS(bottom_pixels),
       CAFFE_CUDA_NUM_THREADS>>>(tmp_data, bottom_shape,
-        mean_values_[0] * UINT8_MAX,
-        mean_values_[1] * UINT8_MAX,
-        mean_values_[2] * UINT8_MAX);
+        mean_values_[0] * __UINT8_MAX__,
+        mean_values_[1] * __UINT8_MAX__,
+        mean_values_[2] * __UINT8_MAX__);
   }
 
   // Spatial transformations
